### PR TITLE
#47483 - Adding NonResourceURLs support to AccessStore

### PR DESCRIPTION
**Pull request by gehrkefc**
This is related to issue: [#47483](https://github.com/rancher/rancher/issues/47483)

### CODE REVIEW

File: pkg/accesscontrol/policy_rule_index.go [hunk 1]
```diff
@@ -85,25 +87,29 @@ func (p *policyRuleIndex) get(subjectName string) *AccessSet {
...
-						accessSet.Add(verb,
-							schema.GroupResource{
-								Group:    group,
-								Resource: resource,
-							}, Access{
-								Namespace:    namespace,
-								ResourceName: resourceName,
-							})
+		if len(rule.Resources) > 0 {
+			for _, group := range rule.APIGroups {
```
Comment (tomleb) [inline pkg/accesscontrol/policy_rule_index.go:91]: Could we perhaps extract some of this to their own methods so we reduce the nesting a bit?

File: pkg/accesscontrol/policy_rule_index.go [hunk 2]
```diff
@@ -85,25 +87,29 @@ func (p *policyRuleIndex) get(subjectName string) *AccessSet {
...
+									Resource: resource,
+								}, Access{
+									Namespace:    namespace,
+									ResourceName: resourceName,
+								})
+						}
 					}
 				}
 			}
+		} else if roleRef.Kind == clusterRoleKind {
```
Comment (tomleb) [inline pkg/accesscontrol/policy_rule_index.go:111]: Should this check that the binding is a clusterrolebinding and not a rolebinding? Seems like for clusterrolebinding we're setting namespace as `All` (which is `*`)

File: pkg/accesscontrol/access_set.go [hunk 3]
```diff
@@ -56,6 +65,13 @@ func (a *AccessSet) Merge(right *AccessSet) {
 			m[k] = v
 		}
 	}
+
+	for k, v := range right.nonResourceSet {
+		_, ok := a.nonResourceSet[k]
+		if !ok {
+			a.nonResourceSet[k] = v
+		}
+	}
```
[the remark this answers is not in the document]
Comment (gehrkefc) [reply pkg/accesscontrol/access_set.go:74]: Ok. Makes sense. I'll add code to merge rules.

File: pkg/accesscontrol/access_set_test.go [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+				assert.Contains(t, rules, tt.expected, "rule not found")
+			} else {
+				assert.NotContains(t, rules, tt.expected, "unexpected rule found")
+			}
+		})
+
+	}
+}
+
+func TestAccessSet_GrantsNonResource(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (gehrkefc) [reply pkg/accesscontrol/access_set_test.go:119]: Ok, I'll add it.

File: pkg/ext/apiserver_authorization_test.go [hunk 5]
```diff
@@ -342,3 +345,152 @@ func (s *ExtensionAPIServerSuite) TestAuthorization() {
 		})
 	}
 }
+
+func TestAuthorization_NonResourceURLs(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (gehrkefc) [reply pkg/ext/apiserver_authorization_test.go:349]: Are these urls being served in any way ?

File: pkg/ext/apiserver_authorization.go [hunk 6]
```diff
@@ -22,20 +22,25 @@ func NewAccessSetAuthorizer(asl accesscontrol.AccessSetLookup) *AccessSetAuthori
...
+	accessSet := a.asl.AccessFor(attrs.GetUser())
+
 	if !attrs.IsResourceRequest() {
-		// XXX: Implement
-		return authorizer.DecisionDeny, "AccessSetAuthorizer does not support nonResourceURLs requests", nil
+		if accessSet.GrantsNonResource(verb, path) {
+			return authorizer.DecisionAllow, "", nil
+		}
+
+		return authorizer.DecisionDeny, "", nil
```
[the remark this answers is not in the document]
Comment (gehrkefc) [reply pkg/ext/apiserver_authorization.go:34]: Ok, Will do.

File: pkg/accesscontrol/policy_rule_index.go (hunk 1 above)
Comment (gehrkefc) [reply pkg/accesscontrol/policy_rule_index.go:91]: I'm sorry, what do you mean exactly ?

File: pkg/ext/apiserver_authorization_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (tomleb) [reply pkg/ext/apiserver_authorization_test.go:349]: Yes these urls are expected to be served to user so they can access the public documentation (OpenAPI spec)

File: pkg/accesscontrol/policy_rule_index.go (hunk 1 above)
Comment (tomleb) [reply pkg/accesscontrol/policy_rule_index.go:91]: Basically, we're doing nested-ifs and nested-loops 6 levels deep which affects the readability. My suggestion would be to split it a little in two functions to make it more readable. eg: the nonResourceURL case and the resourceURL case.

File: pkg/accesscontrol/access_set.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (gehrkefc) [reply pkg/accesscontrol/access_set.go:78]: I changed my mind. Changed it to be `map[nonResourceKey]struct{}`. I've analyzed it and it makes more sense this way. Considering we'll have a look check the rules inside the function from kubernetes.

File: pkg/ext/apiserver_authorization_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (gehrkefc) [reply pkg/ext/apiserver_authorization_test.go:409]: Ok, some bindings using read-only and read-write users (already present) and created 6 new tests.

File: pkg/accesscontrol/policy_rule_index.go (hunk 1 above)
Comment (gehrkefc) [reply pkg/accesscontrol/policy_rule_index.go]: Created another function just to handle resource based rules. It became this way:

https://github.com/rancher/steve/blob/026f994f5803054edb249ac5f2080457b4b50c18/pkg/accesscontrol/policy_rule_index.go#L89-L94

File: pkg/accesscontrol/access_set_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (gehrkefc) [reply pkg/accesscontrol/access_set_test.go:55]: Added.

File: pkg/accesscontrol/policy_rule_index.go (hunk 2 above)
Comment (gehrkefc) [reply pkg/accesscontrol/policy_rule_index.go]: I'm checking if the kind is "ClusterRole" and not setting the namespace, does it mean "All" in k8s terms ?

File: pkg/ext/apiserver_authorization.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (gehrkefc) [reply pkg/ext/apiserver_authorization.go:37]: Checked it integrating with rancher and added a comment with the response message.

File: pkg/accesscontrol/access_set_test.go [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+			verb:   "post",
+			url:    "/healthz",
+			keys:   map[nonResourceKey]struct{}{},
+			expect: false,
+		},
+	}
+
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			accessSet := &AccessSet{nonResourceSet: tt.keys}
```
Comment (tomleb) [inline pkg/accesscontrol/access_set_test.go:128]: nit: WDYT of using the exported method to construct the AccessSet? That way if the implementation changes we most likely won't have to change the test also.

File: pkg/ext/apiserver_authorization_test.go [hunk 8]
```diff
@@ -311,6 +314,66 @@ func (s *ExtensionAPIServerSuite) TestAuthorization() {
 			},
 			expectedStatusCode: http.StatusForbidden,
 		},
+		{
```
Comment (tomleb) [inline pkg/ext/apiserver_authorization_test.go:317]: I find it a bit confusing that `read-only` means read-only but only for all resources AND for openapi v2, but not v3. I think it would be less confusing to make up new users specifically for those cases. eg: a `openapi-v2-only-read` user and a `openapi-read-only` user, or something similar.

**Review by tomleb — COMMENTED** [review]
Looks good, a couple of things for tests but otherwise :+1:

File: pkg/ext/apiserver_authorization_test.go (hunk 8 above)
Comment (gehrkefc) [reply pkg/ext/apiserver_authorization_test.go:317]: Ok, user changed!

File: pkg/accesscontrol/access_set_test.go (hunk 7 above)
Comment (gehrkefc) [reply pkg/accesscontrol/access_set_test.go:128]: Agree, just changed it.

File: pkg/accesscontrol/policy_rule_index.go (hunk 2 above)
Comment (gehrkefc) [reply pkg/accesscontrol/policy_rule_index.go]: I have this test at GrantsNonResource, should I replicate it here ?

File: pkg/accesscontrol/access_set_test.go [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+	}{
+		{
+			name:  "valid case",
+			verbs: []string{"get", "post"},
+			urls:  []string{"/healthz", "/metrics"},
+			want: []nonResourceKey{
+				{"get", "/healthz"},
+				{"get", "/metrics"},
+				{"post", "/healthz"},
+				{"post", "/metrics"},
```
[the remark this answers is not in the document]
Comment (gehrkefc) [reply pkg/accesscontrol/access_set_test.go:24]: Ok, I had this in GrantsNonResource. Added it here too.

File: pkg/accesscontrol/access_set.go [hunk 10]
```diff
@@ -120,6 +160,25 @@ func (a *AccessSet) Add(verb string, gr schema.GroupResource, access Access) {
 	}
 }
 
+func (a *AccessSet) AddNonResouceURLs(verbs, urls []string) {
```
Comment (ericpromislow) [inline pkg/accesscontrol/access_set.go:163]: typo: should be `AddNonResourceURLs`

File: pkg/accesscontrol/access_set_test.go [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+			want:  nil,
+		},
+	}
+
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			accessSet := &AccessSet{}
+			accessSet.AddNonResouceURLs(tt.verbs, tt.urls)
+
+			if len(tt.want) > 0 {
```
Comment (ericpromislow) [inline pkg/accesscontrol/access_set_test.go:62]: Shouldn't there be an `else` part here to assert that `accessSet.nonResourceSet` doesn't contain certain values?  I should add that usually testing a negative doesn't give as much information as testing a positive, so this might not be a good use of time. Just asking

**Review by ericpromislow — CHANGES_REQUESTED** [review]
Just one typo

File: pkg/accesscontrol/access_set.go (hunk 10 above)
Comment (gehrkefc) [reply pkg/accesscontrol/access_set.go]: Just fixed it. Thanks!

File: pkg/accesscontrol/access_set_test.go (hunk 11 above)
Comment (gehrkefc) [reply pkg/accesscontrol/access_set_test.go:62]: Ok, just added it. Thank you!

**Review by tomleb — APPROVED** [review]
LGTM, let's wait after code freeze before merging this

**Review by ericpromislow — APPROVED** [review]
LGTM
